src/python/pants/engine/target_test.py

```diff
@@ -0,0 +1,191 @@
+        if non_haskell_sources:
+            raise ValueError(
+                f"Received non-Haskell sources in {self.alias}: {non_haskell_sources}."
+            )
+
+
+@dataclass(frozen=True)
+class HaskellSourcesResult:
+    snapshot: Snapshot
+
+
+@rule
+async def hydrate_haskell_sources(sources: HaskellSources) -> HaskellSourcesResult:
+    sources.validate_pre_hydration()
+    result = await Get[Snapshot](PathGlobs("*.hs"))
+    sources.validate_post_hydration(result)
+    return HaskellSourcesResult(result)
+
+
+class HaskellField(PluginField):
+    pass
+
+
+class HaskellTarget(Target):
+    alias: ClassVar = "haskell"
+    core_fields: ClassVar = (HaskellGhcExtensions, HaskellSources)
+    plugin_field_type: ClassVar = HaskellField
+
+
+def test_invalid_fields_rejected() -> None:
+    with pytest.raises(ValueError) as exc:
+        HaskellTarget({"invalid_field": True})
+    assert "Unrecognized field `invalid_field=True` for target type `haskell`." in str(exc)
+
+
+def test_get_primitive_field() -> None:
+    extensions = ["GhcExistentialQuantification"]
+    extensions_field = HaskellTarget({HaskellGhcExtensions.alias: extensions}).get(
+        HaskellGhcExtensions
+    )
+    assert extensions_field.raw_value == extensions
+    assert extensions_field.value == extensions
+
+    default_extensions_field = HaskellTarget({}).get(HaskellGhcExtensions)
+    assert default_extensions_field.raw_value is None
+    assert default_extensions_field.value == []
+
+
+def test_get_async_field() -> None:
+    def hydrate_field(
+        *, raw_source_files: List[str], hydrated_source_files: Tuple[str, ...]
+    ) -> HaskellSourcesResult:
+        sources_field = HaskellTarget({HaskellSources.alias: raw_source_files}).get(HaskellSources)
+        assert sources_field.raw_value == raw_source_files
+        result: HaskellSourcesResult = run_rule(
+            hydrate_haskell_sources,
+            rule_args=[sources_field],
+            mock_gets=[
+                MockGet(
+                    product_type=Snapshot,
+                    subject_type=PathGlobs,
+                    mock=lambda _: Snapshot(
+                        directory_digest=EMPTY_DIRECTORY_DIGEST,
+                        files=hydrated_source_files,
+                        dirs=(),
+                    ),
+                )
+            ],
+        )
+        return result
+
+    # Normal field
+    expected_files = ("monad.hs", "abstract_art.hs", "abstract_algebra.hs")
+    assert (
+        hydrate_field(
+            raw_source_files=["monad.hs", "abstract_*.hs"], hydrated_source_files=expected_files
+        ).snapshot.files
+        == expected_files
+    )
+
+    # Test pre-hydration validation
+    with pytest.raises(ValueError) as exc:
+        hydrate_field(raw_source_files=[0, 1, 2], hydrated_source_files=())  # type: ignore[call-arg]
+    assert "Not all elements of the iterable have type" in str(exc)
```

Comment:
Again user-facing error-wise ideally with will be a lot more guiding; I'd hope it would read something like:
"`java_library` target `some:target` defined in `some/BUILD:15` has an invalid `sources` argument. Expected a list with all elements of type string, but found number element(s)"

Fine with this as-is today, but we should make sure we have a path to something like the above

Reply:
Agreed. There's a path for this. The first change is (likely) storing an `Address` on each `Field` so that they can refer to it in any error messages.